File: java/v3/src/main/java/com/kmeans/Kmeans.java
package com.kmeans;

import java.util.ArrayList;
import java.util.List;

public class Kmeans {
    
    private final int numClusters;
    private final Point[] points;
    private int[] assignments;
    private Point[] centroids;
    private final int numPoints;
    private final double epsilon;
    
    public Kmeans(int numClusters, List<Point> points, long seed) {
        if(points == null || points.size() < numClusters){
            throw new IllegalArgumentException("Number of clusters greater than the number of points");
        }

        this.numClusters = numClusters;
        this.epsilon = 1e-10;
        this.numPoints = points.size();
        this.points = new Point[this.numPoints];
        for (int i = 0; i < this.numPoints; i++) {
            this.points[i] = points.get(i);
        }
        this.assignments = new int[this.numPoints];
        this.centroids = new Point[this.numClusters];
        
        int[] indices = {10, 532, 9012};
        for (int i = 0; i < this.numClusters; i++) {
            this.centroids[i] = new Point(this.points[indices[i]]);
        }
    }

    private int findNearestCentroid(Point p) {
        double minDistance = Double.MAX_VALUE;
        int nearestIndex = -1;
        
        for (int i = 0; i < centroids.length; i++) {
            double dist = p.sqDistanceTo(centroids[i]);
            if (dist < minDistance) {
                minDistance = dist;
                nearestIndex = i;
            }
        }
        return nearestIndex;
    }

    private boolean centroidsConverged(Point[] last, Point[] now) {
        for (int i = 0; i < this.numClusters; i++) {
            if (last[i].sqDistanceTo(now[i]) > this.epsilon) {
                return false;
            }
        }
        return true;
    }

    private void updateCentroids(int numWorkers)  throws InterruptedException {
        int dim = this.points[0].getDimension();
        double[][][] sumsPartial = new double[numWorkers][numClusters][dim];
        int[][] countsPartial = new int[numWorkers][numClusters];
        Thread[] threads = new Thread[numWorkers];
        int grainSize = (this.numPoints + numWorkers - 1) / numWorkers; 

        for (int i = 0; i < numWorkers; i++) {
            final int workerId = i; 
            final int start = i * grainSize;
            final int end = Math.min(start + grainSize, this.numPoints);

            if (start >= this.numPoints) break;

            threads[i] = Thread.ofVirtual().start(() -> {
                for (int j = start; j < end; j++) {
                    int clusterId = this.assignments[j];
                    double[] point = this.points[j].getCoordinates();

                    for (int d = 0; d < dim; d++) {
                        sumsPartial[workerId][clusterId][d] += point[d];
                    }
                    countsPartial[workerId][clusterId]++;
                }
            });
        }

        for (Thread t : threads) {
            while (t.isAlive()){}
        }

        for (int c = 0; c < numClusters; c++) {
            double[] finalSum = new double[dim];
            int totalCount = 0;
            
            for (int w = 0; w < numWorkers; w++) {
                totalCount += countsPartial[w][c];
                for (int d = 0; d < dim; d++) {
                    finalSum[d] += sumsPartial[w][c][d];
                }
            }

            if (totalCount > 0) {
                for (int d = 0; d < dim; d++) {
                    finalSum[d] /= totalCount;
                }
                this.centroids[c] = new Point(finalSum);
            }
        }
    }

    private void clustering(int numWorkers) throws InterruptedException {
        Thread[] threads = new Thread[numWorkers];
        int grainSize = (this.numPoints + numWorkers - 1) / numWorkers; 

        for (int i = 0; i < numWorkers; i++) {
            final int start = i * grainSize;
            final int end = Math.min(start + grainSize, this.numPoints);

            if (start >= this.numPoints) break;

            threads[i] = Thread.ofVirtual().start(() -> {
                for (int j = start; j < end; j++) {
                    this.assignments[j] = findNearestCentroid(this.points[j]);
                }
            });
        }

        for (Thread t : threads) while (t.isAlive()){}
    }

    public void fit() {
        this.fit(300, Runtime.getRuntime().availableProcessors());
    }

    public void fit(int maxIterations, int numWorkers) {
        try {
            Point[] lastCentroids;
            int iteration = 0;
            boolean converged = false;

            do {
                lastCentroids = this.centroids.clone();
                this.clustering(numWorkers);
                this.updateCentroids(numWorkers);
                converged = centroidsConverged(lastCentroids, this.centroids);
                iteration++;
            } while (!converged && iteration < maxIterations);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public Point[] getCentroids() {
        return centroids;
    }

    public List<Point>[] getClusters() {
        @SuppressWarnings("unchecked")
        List<Point>[] clusters = (List<Point>[]) new List[this.numClusters];

        for (int i = 0; i < this.numClusters; i++) {
            clusters[i] = new ArrayList<Point>();
        }

        for (int i = 0; i < this.numPoints; i++) {
            clusters[this.assignments[i]].add(this.points[i]);
        }

        return clusters;
    }
}

File: go/v4/kmeans.go
package main

import (
	"errors"
	"math"
	"runtime"
	"sync"
)

type kmeans struct {
	numClusters int
	points []Point
	assignments []int
	centroids []Point
	numPoints int
	epsilon float64
}

func Kmeans(numClusters int, points []Point, seed uint64) (*kmeans, error) {
	numPoints := len(points)
	if numClusters == 0 || numClusters > numPoints {
		return nil, errors.New("number of clusters greater than the number of points or zero points")
	}

	instance := &kmeans{
		numClusters: numClusters,
		epsilon: 1e-10,
		numPoints: len(points),
		points: points,
		assignments: make([]int, numPoints),
		centroids: make([]Point, numClusters),
	}
	
	indices := []int{0, 1, 2}
	instance.initCentroids(indices)

	return instance, nil
}

func (k *kmeans) initCentroids(indices []int) {
	for i := 0; i < k.numClusters; i++ {
		k.centroids[i] = k.points[indices[i]]
	}
}

func (k *kmeans) findNearestCentroid(p *Point) (int, error) {
	minDistance := math.MaxFloat64
	nearestIndex := -1

	for i := 0; i < k.numClusters; i++ {
		dist, err := p.SqDistanceTo(&k.centroids[i])

		if err != nil {
			return -1, err
		}

		if dist < minDistance {
			minDistance = dist
			nearestIndex = i
		}
	}

	return nearestIndex, nil
}

func (k *kmeans) centroidsConverged(last []Point, now []Point) (bool, error) {
	for i := 0; i < k.numClusters; i++ {
		dist, err := last[i].SqDistanceTo(&now[i])
		if err != nil {
			return false, err
		}
		if  dist > k.epsilon {
			return false, nil
		}
	}
	return true, nil
}

func (k *kmeans) updateCentroids(numWorkers int) error {
	dim := k.points[0].Dimension()

	sumsPartial := make([][][]float64, numWorkers)
	for i := range sumsPartial {
	    sumsPartial[i] = make([][]float64, k.numClusters)
	    for j := range sumsPartial[i] {
	        sumsPartial[i][j] = make([]float64, dim)
	    }
	}
	
	countsPartial := make([][]int, numWorkers)
	for i := range countsPartial {
	    countsPartial[i] = make([]int, k.numClusters)
	}

	var wg sync.WaitGroup
	grainSize := (k.numPoints + numWorkers -1) / numWorkers

	for i := 0; i < numWorkers; i++ {
		workerId := i
		start := i * grainSize
		end := int(math.Min(float64(start + grainSize), float64(k.numPoints)))

		if start >= k.numPoints {
			break
		}

		wg.Add(1)
		go func (workerId int, start int, end int) {
			defer wg.Done()
			runtime.LockOSThread()
        	defer runtime.UnlockOSThread()
			
			for j := start; j < end; j++ {
				clusterId := k.assignments[j]
				point := k.points[j].Coordinates()

				for d := 0; d < dim; d++ {
					sumsPartial[workerId][clusterId][d] += point[d]
				}
				countsPartial[workerId][clusterId]++
			}
		}(workerId, start, end)
	}
	
	wg.Wait()

	for c := 0; c < k.numClusters; c++ {
		finalSum := make([]float64, dim)
		totalCount := 0

		for w := 0; w < numWorkers; w++ {
			totalCount += countsPartial[w][c]
			for d := 0; d < dim; d++ {
				finalSum[d] += sumsPartial[w][c][d]
			}
		}

		if totalCount > 0 {
			for d := 0; d < dim; d++ {
				finalSum[d] /= float64(totalCount)
			}

			point, err := NewPoint(finalSum)

			if err != nil {
				return err
			}

			k.centroids[c] = *point
		}
	}

	return nil
}

func (k *kmeans) clustering(numWorkers int) error {
	grainSize := (k.numPoints + numWorkers - 1) / numWorkers
	var wg sync.WaitGroup
	errChan := make(chan error, numWorkers)

	for i := 0; i < numWorkers; i++ {
		start := i * grainSize
		end := int(math.Min(float64(start+grainSize), float64(k.numPoints)))

		if start >= k.numPoints {
			break
		}

		wg.Add(1)
		go func (start int, end int) {
			defer wg.Done()
			runtime.LockOSThread()
        	defer runtime.UnlockOSThread()

			for j := start; j < end; j++ {
				idx, err := k.findNearestCentroid(&k.points[j])
                if err != nil {
                    select {
                    case errChan <- err:
                    default:
                    }
                    return
                }
                k.assignments[j] = idx
			}
		}(start, end)
	}

	wg.Wait()

	close(errChan)

	if len(errChan) > 0 {
	    return <-errChan
	}

	return nil
}

func (k *kmeans) Fit() error {
	err := k.FitWithArgs(300, runtime.NumCPU())
	
	if err != nil {
		return err
	}

	return nil
}

func (k *kmeans) FitWithArgs(maxIterations int, numWorkers int) error {
	lastCentroids := make([]Point, k.numClusters)
	iteration := 0
	copy(lastCentroids, k.centroids)
	k.clustering(numWorkers)
	k.updateCentroids(numWorkers)
	converged, err := k.centroidsConverged(lastCentroids, k.centroids)
	if err != nil {
		return err
	}
	iteration++

	for {
		if converged || iteration >= maxIterations {
			break
		}

		copy(lastCentroids, k.centroids)
		k.clustering(numWorkers)
		k.updateCentroids(numWorkers)
		converged, err = k.centroidsConverged(lastCentroids, k.centroids)
		if err != nil {
			return err
		}
		iteration++
	}
	return nil
}

func (k *kmeans) Centroids() []Point {
	return k.centroids
}

func (k *kmeans) Clusters() [][]Point {
	clusters := make([][]Point, k.numClusters)

	for i := 0; i < k.numClusters; i++ {
		clusters[i] = make([]Point, 0)
	}

	for i := 0; i < k.numPoints; i++ {
        clusterID := k.assignments[i]
        clusters[clusterID] = append(clusters[clusterID], k.points[i])
    }

	return clusters
}
